java/src/jmri/jmrit/logixng/actions/ActionListenOnBeans.java

```diff
@@ -167,20 +169,41 @@ public void execute() {
         // changes of the registered beans and execute the ConditionalNG
         // when it happens.
 
+        boolean isQueueEmpty;
+
         synchronized(this) {
+            String namedBean;
+            String event;
+            String newValue;
+
+            PropertyChangeEvent evt = _eventQueue.poll();
+            if (evt != null) {
+                namedBean = ((NamedBean)evt.getSource()).getDisplayName();
+                event = evt.getPropertyName();
+                newValue = evt.getNewValue() != null ? evt.getNewValue().toString() : null;
+            } else {
+                namedBean = null;
+                event = null;
+                newValue = null;
+            }
+
             SymbolTable symbolTable = getConditionalNG().getSymbolTable();
+
             if (_localVariableNamedBean != null) {
-                symbolTable.setValue(_localVariableNamedBean, _lastNamedBean);
+                symbolTable.setValue(_localVariableNamedBean, namedBean);
             }
             if (_localVariableEvent != null) {
-                symbolTable.setValue(_localVariableEvent, _lastEvent);
+                symbolTable.setValue(_localVariableEvent, event);
             }
             if (_localVariableNewValue != null) {
-                symbolTable.setValue(_localVariableNewValue, _lastNewValue);
+                symbolTable.setValue(_localVariableNewValue, newValue);
             }
-            _lastNamedBean = null;
-            _lastEvent = null;
-            _lastNewValue = null;
+
+            isQueueEmpty = _eventQueue.isEmpty();
+        }
+
```

Comment:
I'm curious about why you use an `isQueueEmpty` boolean rather than directly checking `_eventUque.isEmpty()` in the `if` statement.  It seems to me that there's still a tiny window between the `isQueueEmpty = _eventQueue.isEmpty();` and the `if (!isQueueEmpty)` where the queue could become non-empty again and the code wouldn't notice.  Or is that not a problem?

Reply:
Thank you! The reason was that I didn't want to have the `getConditionalNG().execute();` statement within the synchronized section to avoid potential deadlooks. But I don't think that's a problem.

And you are right. I need to have the if statement within the synchronized block to prevent the queue to be non empty without noticing it.

I have fixed it now.